### PR TITLE
[Android] Fix onKeyPress so it provides the correct key when TextInput is at max length

### LINKED ISSUE #30747: On Android, the onKeyPress callback returns the wrong key value when the TextInput has the maxLength property
## Description
In Android, when a TextInput with the maxLength property is "full"(that is, the length of the input is equal to the maximum), the onKeyPress event will return the last character in the input.

## React Native version:
0.63.3

## Steps To Reproduce

1. Create a TextInput component with maxLength=1, and an onKeyPress handler
2. Type multiple different characters into the TextInput
3. Observe that event.nativeEvent.key in the onKeyPress callback is always the first character that was entered into the TextInput

## Expected Results
onKeyPress should return the key that was actually pressed.

This behaviour only happens in Android, and can be seen in the following snack:
https://snack.expo.io/PuyXgucEf

### PR DESCRIPTION
**Pull request by osamaqarem**
## Summary

Fixes #30747 making Android behave the same as iOS under the same scenario.

When autosuggest is enabled (which is the default), then the keypress is inferred based on cursor position. The issue appears when max length is reached where we can no longer guess the keypress based on cursor location alone.

The problem can be reproduced using the expo snack provided in the original issue: 
https://snack.expo.io/PuyXgucEf

## Changelog

[Android] [Fixed] -  Fix onKeyPress so it provides the correct key when TextInput has keybard suggestions and is at max length.

## Test Plan

Tested using RNTester app in Components > TextInput > Event handling example and adding a `maxLength` prop.

1. Enter text in the input until `maxLength` is reached.
2. The value printed 'onKeyPress key' should be the actual key pressed.

### Before

https://user-images.githubusercontent.com/24393922/104839131-376eab00-58fa-11eb-8bde-eaddefecd467.mov

### After

https://user-images.githubusercontent.com/24393922/104839187-81f02780-58fa-11eb-9a77-91189d4edefc.mov

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.java [hunk 1]
```diff
@@ -94,10 +94,17 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
     boolean noPreviousSelection = previousSelectionStart == previousSelectionEnd;
     boolean cursorDidNotMove = currentSelectionStart == previousSelectionStart;
     boolean cursorMovedBackwardsOrAtBeginningOfInput =
-        (currentSelectionStart < previousSelectionStart) || currentSelectionStart <= 0;
-    if (cursorMovedBackwardsOrAtBeginningOfInput || (!noPreviousSelection && cursorDidNotMove)) {
+      (currentSelectionStart < previousSelectionStart) || currentSelectionStart <= 0;
+    int maxLength = mEditText.getLengthFilterValue();
```
Comment (osamaqarem) [inline ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.java:99]: Another way of getting the maxLength would be looping through the `InputFilters` on the `EditText` until we find the `InputLengthFilter`. Let me know if that approach is preferrable.

File: ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.java [hunk 2]
```diff
@@ -57,6 +57,7 @@
   private EventDispatcher mEventDispatcher;
   private boolean mIsBatchEdit;
   private @Nullable String mKey = null;
+  private int previousComposingTextLength;
```
Comment (osamaqarem) [inline ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.java:60]: This commit fixes an edge case issue introduced by the [previous commit](https://github.com/facebook/react-native/pull/30751/commits/dfe940205bbb7a10c93a7a34afdd3e4a53855fcd). 

When `maxLength` is set to 1 or more and the max length is reached, the composing text never changes. So backspace can be inferred correctly. However, when `maxLength` is set to 0 then whatever the user types is appended to the composing text and therefore we can no longer infer the backspace keypress.

The videos below demonstrate the issue and the result of the fix introduced in this commit.
In both theses videos, `maxLength` is set to 0.

### Before


https://user-images.githubusercontent.com/24393922/104837671-105fab80-58f1-11eb-9a7f-c293299e5a54.mov


### After


https://user-images.githubusercontent.com/24393922/104837919-c4ae0180-58f2-11eb-81dc-fd199357238b.mov

**Comment by osamaqarem** [comment]
Regarding the commit: [fix: backspace keypress not inferred when max length is 0](https://github.com/facebook/react-native/pull/30751/commits/bb90b76b99b1b27a6d49d8fd69fe9c6b3aae670f):

It fixes an edge case issue introduced by the [previous commit](https://github.com/facebook/react-native/pull/30751/commits/dfe940205bbb7a10c93a7a34afdd3e4a53855fcd). 

When `maxLength` is set to 1 or more and the max length is reached, the composing text never changes. So backspace can be inferred correctly. However, when `maxLength` is set to 0 then whatever the user types is appended to the composing text and therefore we can no longer infer the backspace keypress.

The videos below demonstrate the issue and the result of the fix introduced in this commit.
In both theses videos, `maxLength` is set to 0.

### Before


https://user-images.githubusercontent.com/24393922/104837671-105fab80-58f1-11eb-9a7f-c293299e5a54.mov


### After


https://user-images.githubusercontent.com/24393922/104837919-c4ae0180-58f2-11eb-81dc-fd199357238b.mov

**Comment by safaiyeh** [comment]
cc @JoshuaGross to check out this PR.

Quick turn around for a reported issue here: https://github.com/facebook/react-native/issues/30747

File: ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java [hunk 3]
```diff
@@ -109,6 +109,7 @@
   private @Nullable String mFontFamily = null;
   private int mFontWeight = ReactTypefaceUtils.UNSET;
   private int mFontStyle = ReactTypefaceUtils.UNSET;
+  private int lengthFilterValue = UNSET;
```
Comment (PrayaAmadigaPitasa) [inline ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java:112]:
```suggestion
  private int mLengthFilterValue = UNSET;
```

Based on the previous code style, react-native using "m" for non public and non static field.

File: ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.java (hunk 2 above)
Comment (PrayaAmadigaPitasa) [inline ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditTextInputConnectionWrapper.java:60]:
```suggestion
  private int mPreviousComposingTextLength;
```

**Pull request closed without merging**